remount/remount.c

```diff
@@ -0,0 +1,605 @@
+{
+	printf("DEBUG: parsing remount object...\n\r");
+
+	int tokc;
+	jsmntok_t *tokv = NULL;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0) {
+		printf("WARN: JSON could not be parsed\n\r");
+		goto out;
+	}
+
+	char *policy = NULL;
+	policy = getenv("PV_REMOUNT_POLICY");
+	if (!policy)
+		policy = getenv("pv_remount_policy");
+	if (!policy)
+		policy = "default";
+	printf("DEBUG: remount policy '%s' to be used\n\r", policy);
+
+	char *remount_array = NULL;
+	remount_array = pv_json_get_value(json, policy, tokv, tokc);
+	if (!remount_array) {
+		printf("WARN: '%s' key not found\n\r", policy);
+		goto out;
+	}
+
+	parse_remount_array(remount_array, remount_entries);
+	free(remount_array);
+
+out:
+	if (tokv)
+		free(tokv);
+}
+
+void parse_device_run_object(char *json, struct dl_list *remount_entries)
+{
+	printf("DEBUG: parsing component JSON...\n\r");
+
+	int tokc;
+	jsmntok_t *tokv = NULL;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0) {
+		printf("WARN: JSON could not be parsed\n\r");
+		goto out;
+	}
+
+	char *remount_object = NULL;
+	remount_object = pv_json_get_value(json, "remount", tokv, tokc);
+	if (!remount_object) {
+		printf("WARN: remount key not found\n\r");
+		goto out;
+	}
+
+	parse_remount_object(remount_object, remount_entries);
+	free(remount_object);
+
+out:
+	if (tokv)
+		free(tokv);
+}
+
+char *get_run_json_rel_path(void)
+{
+	char *pname = NULL;
+	pname = getenv("LXC_NAME");
+	if (!pname) {
+		printf("ERROR: LXC_NAME env not set\n\r");
+		return NULL;
+	}
+	printf("DEBUG: container name '%s' to be used\n\r", pname);
+
+	char *path_fmt = "%s/run.json";
+	size_t len = strlen(path_fmt) + strlen(pname) + 1;
+	char *path = calloc(1, len);
+	snprintf(path, len, path_fmt, pname);
+	return path;
+}
+
+int parse_state_json(char *json, struct dl_list *remount_entries)
+{
+	printf("DEBUG: parsing state JSON...\n\r");
+
+	int tokc, ret = -1;
+	jsmntok_t *tokv = NULL;
+
+	if (jsmnutil_parse_json(json, &tokv, &tokc) < 0) {
+		printf("WARN: JSON could not be parsed\n\r");
+		goto out;
+	}
+
+	printf("DEBUG: searching for device.json...\n\r");
+
+	char *device_json = NULL;
+	device_json = pv_json_get_value(json, "device.json", tokv, tokc);
+	if (device_json) {
+		parse_device_run_object(device_json, remount_entries);
+		free(device_json);
+	} else
+		printf("WARN: device.json not found\n\r");
+
+	char *rel_path = NULL;
+	rel_path = get_run_json_rel_path();
+	if (!rel_path) {
+		printf("ERROR: could not form run.json relative path\n\r");
+		goto out;
+	}
+
+	printf("DEBUG: searching for '%s'...\n\r", rel_path);
+
+	char *run_json = NULL;
+	run_json = pv_json_get_value(json, rel_path, tokv, tokc);
+	if (run_json) {
+		parse_device_run_object(run_json, remount_entries);
+		free(run_json);
+	} else
+		printf("WARN: '%s' not found\n\r", rel_path);
+
+	ret = 0;
+
+out:
+	if (rel_path)
+		free(rel_path);
+	if (tokv)
+		free(tokv);
+
+	return ret;
+}
+
+#define STORAGE_TRAILS_CURRENT_PVR_JSON_PATH "/storage/trails/current/.pvr/json"
```

Comment:
Should this type of constants be placed at the beginning of the file?

Reply:
done